cloudapi/insights/domain.go

```diff
@@ -0,0 +1,103 @@
+package insights
+
+import (
+	"errors"
+	"time"
+)
+
+var (
+	errEmptyTraceID                      = errors.New("empty trace id")
+	errEmptyStartTime                    = errors.New("empty start time")
+	errEmptyEndTime                      = errors.New("empty end time")
+	errEmptyTestRunID                    = errors.New("empty test run id")
+	errEmptyTestRunScenario              = errors.New("empty test run scenario")
+	errEmptyTestRunGroup                 = errors.New("empty test run group")
+	errEmptyProtocolLabels               = errors.New("empty protocol labels")
+	errEmptyProtocolHTTPLabelsURL        = errors.New("empty protocol http labels url")
+	errEmptyProtocolHTTPLabelsMethod     = errors.New("empty protocol http labels method")
+	errEmptyProtocolHTTPLabelsStatusCode = errors.New("empty protocol http labels status code")
+)
+
+// TestRunLabels describes labels associated with a single test run.
+type TestRunLabels struct {
+	ID       int64
+	Scenario string
+	Group    string
+}
+
+// ProtocolLabels is a dummy interface that is used for compile-time type checking.
+type ProtocolLabels interface {
+	IsProtocolLabels()
+}
+
+// ProtocolHTTPLabels describes labels associated with a single HTTP request.
+type ProtocolHTTPLabels struct {
+	URL        string
+	Method     string
+	StatusCode int64
+}
+
+// IsProtocolLabels is a dummy implementation to satisfy the ProtocolLabels interface.
+func (ProtocolHTTPLabels) IsProtocolLabels() {
+	// Do nothing
+}
+
+// RequestMetadatas is a slice of RequestMetadata.
+type RequestMetadatas []RequestMetadata
+
+// RequestMetadata describes metadata associated with a single *traced* request.
+type RequestMetadata struct {
+	TraceID        string
+	Start          time.Time
+	End            time.Time
+	TestRunLabels  TestRunLabels
+	ProtocolLabels ProtocolLabels
+}
+
+// Valid returns an error if the RequestMetadata is invalid.
+// The RequestMetadata is considered invalid if any of the
+// fields zero-value or if the ProtocolLabels type is invalid.
+func (rm RequestMetadata) Valid() error {
+	if rm.TraceID == "" {
+		return errEmptyTraceID
+	}
+
+	if rm.Start.IsZero() {
+		return errEmptyStartTime
+	}
+
+	if rm.End.IsZero() {
+		return errEmptyEndTime
+	}
+
+	if rm.TestRunLabels.ID == 0 {
+		return errEmptyTestRunID
+	}
+
+	if rm.TestRunLabels.Scenario == "" {
+		return errEmptyTestRunScenario
+	}
+
+	if rm.TestRunLabels.Group == "" {
+		return errEmptyTestRunGroup
+	}
```

Comment:
All the values must be present, this is how our backend works. The other part of the code should fill these gaps with "unknown" if not set.

Reply:
> The other part of the code should fill these gaps with "unknown" if not set.

But `group` can have (and does) an empty value for metrics emitted outside of a group. 

So this validation (which is now removed) would've dropped all traces for request done outside of a group.

And if you do that on backside you will also do it.

The function that sets it to `unknown` only does it if it is *missing* not if the value is empty.

Which ... seems just as well to in practice be done on the backend. If you see an empty `group` - set it to unknwon there. Instead of setting it to `unknown` from here.

Specifically here `unknown` kind of ... "not allowed" value for `group` ... but even if it was I would not make a "magical" string be what I mean when a field is missing. The empty value seems just as well, especially if you don't want to differntiate between it and the thing not being set.